addons/udes_advanced_packaging/models/stock_quant_package.py

```diff
@@ -0,0 +1,100 @@
+"""
+Package model: advanced features.
+
+The term package has two meanings:
+- a generic container, either a package or a package
+- a container containing only quants
+
+A pallet is a containers which contains only quants.
+
+A container may not contain a mix of packages and quants.
+"""
+import logging
+import re
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+class StockQuantPackage(models.Model):
+    """A package contains one or more quants or subpackages."""
+
+    _inherit = ["stock.quant.package", "mixin.stock.model"]
+    _name = "stock.quant.package"
+    _sql_constraints = [("name_unique", "unique (name)", "This name has already been used")]
+
+    @api.constrains("name")
+    def _check_name(self):
+        User = self.env["res.users"]
+
+        warehouse = User.get_user_warehouse()
+        self.ensure_one()
+        # Prevent renaming pallet to package if pallet contains packages.
+        if self.search_count([("parent_id", "=", self.id)]) > 0:
```

Comment:
Sorry I didn't spot this one earlier, but the same thing applies (extracting out `Package = self.env["stock.quant.package"]` and doing `Package.search_count(...)`

Reply:
Fixed in 7252a2b, please review.